docs/preview/yamlgettingstarted-templates.md

```diff
@@ -0,0 +1,295 @@
+# YAML getting started - Templates (not available, for discussion only)
+
+## Example: Simple input
+
+The following example illustrates two things:
+
+1. Front-matter is used to define the default values for inputs.
+
+2. Inputs are expanded using template expressions. A value that starts and ends
+   with `${{ }}` indicates a template expression.
+
+Template:
+
+```yaml
+---
+# Default values
+inputs:
+  solution: '**/*.sln'
+---
+steps:
+- task: msbuild@1
+  inputs:
+    solution: ${{ inputs.solution }}
+- task: vstest@2
+  inputs:
+    solution: ${{ inputs.solution }}
+```
+
+Consumer:
+
+```yaml
+steps:
+- template: steps/msbuild.yml
+  inputs:
+    solution: my.sln
+```
+
+## Example: Insert into an array
+
+Template:
+
+```yaml
+---
+# Default values
+inputs:
+  preBuild: []
+  postBuild: []
+---
+phases:
+- phase: build
+  steps:
+  - ${{ inputs.preBuild }}
+  - task: msbuild@1
+  - task: vstest@2
+  - ${{ inputs.postBuild }}
+```
+
+Consumer:
+
+```yaml
+phases:
+- template: phases/build.yml
+  inputs:
+    preBuild:
+    - script: echo hello from pre build
+    postBuild:
+    - script: echo hello from post build
+```
+
+Note, when an array is inserted into an array, the nested array is flattened.
+
+## Example: Insert into a dictionary
+
+Although YAML has a syntax for inserting into a dictionary, we need to invent
+something because we are post-processing the file.
+
+In the example below, the `${{ insert }}` property indicates the value that follows
+is expected to be a mapping, and should be inserted into the outer mapping.
+
+Template:
+
+```yaml
+---
+# Default values
+inputs:
+  variables: {}
+---
+phases:
+- phase: build
+  variables:
+    configuration: debug
+    arch: x86
+    ${{ insert }}: ${{ inputs.variables }}
```

Comment:
For inserting into dictionaries, how about `${{ key }}` instead of `${{ insert }}`?  I.e. `key: value`

Reply:
I thought had discussed not requiring the explicit insert if you are simply merging a list or a dictionary.